runtime/parachains/src/disputes/migration.rs

```diff
@@ -0,0 +1,61 @@
+// Copyright 2017-2022 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Storage migration(s) related to disputes pallet
+
+#[allow(deprecated)]
+use crate::disputes::{pallet::SpamSlots, Config, Pallet};
+use frame_support::{
+	pallet_prelude::*,
+	traits::{OnRuntimeUpgrade, StorageVersion},
+	weights::Weight,
+};
+/// The current storage version.
+const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+pub struct MigrateToV1<T>(sp_std::marker::PhantomData<T>);
+impl<T: Config> OnRuntimeUpgrade for MigrateToV1<T> {
+	fn on_runtime_upgrade() -> Weight {
+		let mut weight: Weight = Weight::zero();
+
+		if StorageVersion::get::<Pallet<T>>() < STORAGE_VERSION {
+			weight += migrate_to_v1::<T>();
+			STORAGE_VERSION.put::<Pallet<T>>();
+		}
+		weight
+	}
+}
+
+/// Migrates the pallet storage to the most recent version, checking and setting the `StorageVersion`.
+pub fn migrate_to_v1<T: Config>() -> Weight {
+	let weight: Weight = Weight::zero();
+
+	// SpamSlots should not contain too many keys so removing everything at once should be safe
+	loop {
```

Comment:
This is something which bothers me a lot. I understand I can't just remove a lot of data from the runtime because it might be an expensive operation. How this should be handled in general?

In this particular case I don't expect to have a lot of keys in `SpamSlots` so this is probably okay. But I think the migration code should safely handle the case where there actually is a lot of data.

Reply:
Good question. @kianenigma can you chime in here as well please? Finding [this](https://github.com/paritytech/substrate/issues/7911) multi block migrations seem to be messy. Can we safely punt on this? In particular what would happen in the worst case if this was large?

Reply:
> In this particular case I don't expect to have a lot of keys in SpamSlots so this is probably okay.

We can measure this by looking at the state, then let's leave it up to chance or spculation.

You should have a good idea of how many keys you are deleting here. 

If few hundred to a few thousand max, then fine within a block. If more, we can speak about how to do it multi-block.

Reply:
Thanks @kianenigma ! I am actually only worried about an attack. Spamslots should be empty right now, but they could get filled up on purpose by an attacker when this hits production. Looking at the state does not really help. 

What would be the worst an attacker could cause here? An alternative could also be to make this two stages:

1. Release a runtime which ignores and deprecates spam slots (does not fill them up any more).
2. Make a second release with a migration actually removing them, which would then be safe by state checking before merging.

Reply:
I guess this makes sense. @tdimitrov let's just factor out the actual removal into another PR and just deprecate it in this one (and remove the using code).

Reply:
I agree with Kian [here](https://github.com/paritytech/polkadot/pull/6345#discussion_r1054137294) that we should remove the storage and not just deprecate.

Reply:
> let's just factor out the actual removal into another PR and just deprecate it in this one (and remove the using code).

@eskimor I followed the suggestion [here](https://github.com/paritytech/polkadot/pull/6345#discussion_r1054137294) and removed the `SpamSlots` definition. You are okay with this approach, right?